src/mono/mono/mini/method-to-ir.c

```diff
@@ -7514,6 +7514,89 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+				}
+
+				GArray *new_params = g_array_new (FALSE, FALSE, sizeof (MonoType*));
+				uint32_t new_param_count = 0;
+				MonoClass *swift_self = mono_class_try_get_swift_self_class ();
+				MonoClass *swift_error = mono_class_try_get_swift_error_class ();
+				/*
+				 * Go through the lowered arguments, if the argument is a struct, 
+				 * we need to replace it with a sequence of lowered arguments.
+				 * Also record the updated parameters for the new signature.
+				 */
+				for (int idx_param = 0; idx_param < n; ++idx_param) {
+					MonoType *ptype = fsig->params [idx_param];
+					MonoClass *klass = mono_class_from_mono_type_internal (ptype);
+
+					// SwiftSelf and SwiftError are special cases where we need to preserve the class information for the codegen to handle them correctly.
+					if (mono_type_is_struct (ptype) && !(klass == swift_self || klass == swift_error)) {
+						SwiftPhysicalLowering lowered_swift_struct = mono_marshal_get_swift_physical_lowering (ptype, FALSE);
+						if (!lowered_swift_struct.by_reference) {
+							// Create a new local variable to store the base address of the struct
+							MonoInst *struct_base_address =  mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
+							CHECK_ARG (idx_param);
+							NEW_ARGLOADA (cfg, struct_base_address, idx_param);
+							MONO_ADD_INS (cfg->cbb, struct_base_address);
+
+							for (uint32_t idx_lowered = 0; idx_lowered < lowered_swift_struct.num_lowered_elements; ++idx_lowered) {
+								MonoInst *lowered_arg = NULL;
+								// Load the lowered elements of the struct
+								lowered_arg = mini_emit_memory_load (cfg, lowered_swift_struct.lowered_elements [idx_lowered], struct_base_address, lowered_swift_struct.offsets [idx_lowered], 0);
+								*sp++ = lowered_arg;
+
+								++new_param_count;
+								g_array_append_val (new_params, lowered_swift_struct.lowered_elements [idx_lowered]);
+							}
+						} else {
+							// For structs that cannot be lowered, we change the argument to byref type
+							ptype = mono_class_get_byref_type (mono_defaults.typed_reference_class);
+							// Load the address of the struct
+							MonoInst *struct_base_address =  mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
+							CHECK_ARG (idx_param);
+							NEW_ARGLOADA (cfg, struct_base_address, idx_param);
+							MONO_ADD_INS (cfg->cbb, struct_base_address);
+							*sp++ = struct_base_address;
+
+							++new_param_count;
+							g_array_append_val (new_params, ptype);
+						}
+					} else {
+						// Copy over non-struct arguments
+						*sp++ = old_args [idx_param];
+
+						++new_param_count;
+						g_array_append_val (new_params, ptype);
+					}
+				}
+
+				// Create a new dummy signature with the lowered arguments
+				MonoMethodSignature *swiftcall_signature = mono_metadata_signature_alloc (m_class_get_image (method->klass), new_param_count);
+				for (uint32_t idx_param = 0; idx_param < new_param_count; ++idx_param) {
```

Comment:
Can the new signature be created in the "lowering" loop above?

Reply:
We would have to collect the lowering information before otherwise the final number of parameters is not known until all params are lowered. So we would still be doing essentially two loops.